File: scala/SimpleAk/src/main/scala-2.11/actor/TellArticleParser.scala
package actor

import java.util.concurrent.TimeoutException

import akka.actor.Status.Failure
import akka.actor.{Actor, Props, ActorRef}
import akka.actor.Actor.Receive
import akka.util.Timeout
import mol._

/**
 * Created by cqshinn on 2/13/16.
 */
class TellArticleParser(cacheActorPath: String,
                        httpClientActorPath: String,
                        acticleParserActorPath: String,
                        implicit val timeout: Timeout) extends Actor{
  val cacheActor = context.actorSelection(cacheActorPath)
  val httpClientActor = context.actorSelection(httpClientActorPath)
  val articleParserActor = context.actorSelection(acticleParserActorPath)

  implicit val dispatcher = context.dispatcher

  override def receive: Receive = {
    case mes @ ParseArticle(url) => {
      val extraActor: ActorRef = buildExtraActor(sender(), url)
      cacheActor ! (GetRequest(Some(url)), extraActor)
      httpClientActor ! ("test", extraActor)
      context.system.scheduler.scheduleOnce(timeout.duration, extraActor, "timeout")
    }
  }

  private def buildExtraActor(actor: ActorRef, url: String): ActorRef = {
    return context.actorOf(Props(new Actor {
      override def receive: Actor.Receive = {
        case "timeout" =>
          actor ! Failure(new TimeoutException("timeout"))
          context.stop(self)
        case HttpResponse(body) =>
          articleParserActor ! ParseHtmlArticle(url, body)

        case body: String =>
          actor ! body
          context.stop(self)

        case ArticleBody(uri, body) =>
          cacheActor ! SetRequest(Some(url), Some(body))
          actor ! body
          context.stop(self)
        case t =>
          println("Any")
      }
    }));
  }
}

File: java/SimpleAk/src/main/java/actor/TellArticleParser.java
package actor;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSelection;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;
import akka.util.Timeout;
import mol.*;
import scala.PartialFunction;

import javax.sound.midi.Receiver;
import java.util.concurrent.TimeoutException;

/**
 * Created by cqshinn on 2/13/16.
 */
public class TellArticleParser extends AbstractActor {
    private final ActorSelection cacheActor;
    private final ActorSelection httpClientActor;
    private final ActorSelection artcileParseActor;
    private final Timeout timeout;

    public TellArticleParser(String cacheActorPath, String httpClientActorPath,
                             String artcileParseActorPath, Timeout timeout) {
        this.cacheActor = context().actorSelection(cacheActorPath);
        this.httpClientActor = context().actorSelection(httpClientActorPath);
        this.artcileParseActor = context().actorSelection(artcileParseActorPath);
        this.timeout = timeout;
    }

    public PartialFunction receive(){
        return ReceiveBuilder.match(ParseArticle.class, mes -> {
            ActorRef ref = buildExtraActor(sender(), mes.url);
            cacheActor.tell(new GetRequest(mes.url),ref);
            httpClientActor.tell(mes.url,ref);
            context().system().scheduler().scheduleOnce(timeout.duration(), ref, "timeout",
                    context().system().dispatcher(), ActorRef.noSender());

        }).build();
    }

    private ActorRef buildExtraActor(ActorRef actor, String url){
        class MyActor extends AbstractActor {
            public MyActor(){
                receive(ReceiveBuilder
                                .matchEquals(String.class, mes -> mes.equals("timeout"), mes -> {
                                    actor.tell(new akka.actor.Status.Failure(new TimeoutException("timeout!")), self());
                                    context().stop(self());
                                }).match(HttpResponse.class, mes -> {
                                    artcileParseActor.tell(new ParseHtmlArticle(url, mes.body), self());
                                }).match(String.class, mes -> {
                                    actor.tell(mes, self());
                                    context().stop(self());
                                }).match(ArticleBody.class, mes -> {
                                    cacheActor.tell(new SetRequest(mes.url, mes.body), self());
                                    sender().tell(mes.body, self());
                                    context().stop(self());
                                }).matchAny(mes -> {
                                    System.out.println("Any");
                                })
                                .build()
                );
            }
        }
        return context().actorOf(Props.create(MyActor.class, () -> new MyActor()));

    }
}

